### DISCUSSION TITLE
Reuse template computation

### DISCUSSION THREAD

**Discussion by miquecg** [twpayne/chezmoi#2703] [Q&A]
Let's say I have a small "computation" using some data stored in `chezmoi.toml.tmpl` and this is repeated over several templates.
```
{{ index (. | sortAlpha | reverse) 0 }}
```
I'm curious to know if there would be ways to reuse template code like this. I thought of two possible options:
- Compute the result on `chezmoi.toml.tmpl` with the caveat that it uses data defined there, so I imagine it will not be available during evaluation
- Compute the result in another template and then include it where needed. All data would be available.

For something that small I'm fine with copying/pasting, but I'm wondering how one would solve this.

**Comment by halostatue** [comment]
What I’m going to suggest here is useful only if the `| sortAlpha | reverse` is *expensive* to do, because it’s *not* shorter.

As of the moment, storing the result is your best bet. I’d suggest a `before` script (templated) that does something like:

```bash
#!/bin/bash

cat > /tmp/sortedData <<EOF
{{ .dataToSort | sortAlpha | reverse | toJson }}
EOF
```

Then you should be able do `index ((output "cat" "/tmp/sortedData") | fromJson) 0` in your remaining templates.

At the end, you can have an `after` script that just does `rm -f /tmp/sortedData`.

**Reply by miquecg** [reply]
Sorry, just edited my question. I was thinking that simply using a template as storage mechanism would work.

**Reply by halostatue** [reply]
As @twpayne you could put `{{ . | sortAlpha | reverse }}` in `.chezmoitemplates/reverseAlpha` (I don’t remember if `.tmpl` is required in `.chezmoitemplates`) and do `{{ index (includeTemplate "reverseAlpha" .) 0 }}` so that you have a "function" like template call.

**Comment by halostatue** [comment]
@twpayne @bradenhilton this *might* be a good idea for the #2673 templating. Something like `{{- .dataToSort | sortAlpha | reverse | storeDataAs "key" -}}` and `{{ index (readDataAs "key") 0 }}` to have temporary files (or in-memory storage) associated with a *specific* execution context of Chezmoi (so the files might be done as `/tmp/chezmoi-cache-$PID.json` or something like that, and cleaned up on removal).

**Comment by twpayne** [comment]
You can reuse template code by using [`.chezmoitemplates`](https://www.chezmoi.io/reference/special-files-and-directories/chezmoitemplates/).

There is no way in chezmoi to store the result of a template execution and re-use it elsewhere. Such a thing is both a very bad idea and impossible to implement.

It is impossible to implement because chezmoi's templates are neither hermetic nor idempotent so chezmoi cannot determine when the result of a template execution can be reused. Consider the small "computation" `{{ output "head" "-c1" "/dev/random" }}`.

It is a very bad idea because, even it were possible to implement correctly (which would first require solving the halting problem) it would result in a large, shared, global state, with all the nightmares that that brings. Think how bad global variables are. This would be even worse.

**Comment by miquecg** [comment]
Thanks both for engaging :) So basically that's option 2, reuse template code by including them. The word "compute" could be a bit misleading, but in reality I was talking about code reuse.
